### PR TITLE
add init params for Ellipse & Circle and RANSAC

### PR DESCRIPTION
**Pull request by Borda**
## Description
adding option to use some init parameters for model fitting, eg, you have some a priory knowledge, you do not want to use some default parameters

## Checklist
[It's fine to submit PRs which are a work in progress! But before they are merged, all PRs should provide:]
- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- [x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [x] Gallery example in `./doc/examples` (new features only)
- [x] Unit tests

[For detailed information on these and other aspects see [scikit-image contribution guidelines](http://scikit-image.org/docs/dev/contribute.html)]

### CODE REVIEW

**Comment by Borda** [comment]
just example on real data shows that whether we have a priory knowledge the result could be better or at leas closer our expectation

```
center = [241.7921398551373, 696.9405794509523]
points = np.array(
    [[105, 835], [105, 845], [105, 865], [105, 875], [107, 855], [115, 825],
    [115, 835], [116, 845], [115, 865], [115, 885], [117, 855], [125, 825],
    [125, 835], [125, 895], [135, 818], [134, 893], [137, 888], [144, 832],
    [146, 825], [144, 895], [141, 826], [155, 897], [155, 675], [156, 695],
    [155, 825], [157, 685], [153, 831], [163, 726], [164, 824], [165, 902],
    [165, 645], [166, 655], [165, 665], [164, 705], [164, 715], [166, 735],
    [166, 745], [166, 674], [165, 695], [167, 684], [173, 724], [173, 705],
    [175, 715], [175, 818], [175, 635], [175, 645], [174, 755], [175, 902],
    [176, 655], [177, 665], [176, 735], [177, 745], [177, 765], [185, 754],
    [184, 885], [184, 635], [186, 822], [186, 895], [185, 774], [182, 893],
    [188, 763], [194, 884], [193, 777], [196, 768], [195, 785], [195, 817],
    [192, 881], [203, 824], [205, 876], [204, 777], [203, 885], [200, 781],
    [204, 788], [214, 783], [215, 794], [215, 892], [216, 816], [221, 876],
    [223, 825], [225, 886], [225, 784], [225, 795], [225, 805], [227, 834],
    [234, 793], [230, 880], [235, 883], [245, 875], [245, 605], [245, 801],
    [240, 799], [244, 814], [247, 826], [255, 796], [253, 864], [255, 605],
    [251, 860], [265, 607], [274, 606], [275, 596], [285, 597], [284, 607],
    [295, 605], [294, 616], [295, 625], [297, 764], [304, 615], [303, 635],
    [305, 625], [305, 647], [306, 656], [307, 715], [306, 744], [307, 762],
    [313, 635], [313, 645], [312, 734], [314, 756], [316, 767], [312, 665],
    [313, 684], [313, 695], [317, 654], [314, 705], [315, 745], [315, 725],
    [318, 674], [317, 715], [323, 665], [323, 695], [322, 735], [323, 685],
    [325, 705], [325, 763], [325, 725], [327, 675]])

ransac_model, inliers = ransac(points, EllipseModel, min_samples=0.4,
                               residual_threshold=15, max_trials=50)
print 'inliers: ', np.sum(inliers)

dists = spatial.distance.cdist(points, [center], metric='euclidean')
close = (dists <= np.mean(dists))[:, 0]
ransac_model, inliers = ransac(points, EllipseModel, min_samples=0.4,
                               residual_threshold=15, max_trials=50, init_samples=close)
print 'inliers: ', np.sum(inliers)
```
 gives following results...

`inliers:  68`
`inliers:  85`

**Comment by sciunto** [comment]
@Borda What's the comparison with the algorithm you mentioned in #2379?

**Comment by Borda** [comment]
@sciunto here is the previous algorithm, not the mentioned in #2379 but if the change is agreed, I am fine to implement it as part of this PR (or create new PR just for the ellipse?)

**Comment by sciunto** [comment]
@Borda I think it's better to create a new PR if you change the implementation ;) 

Actually, I was just wondering if this patch could improve the situation you exposed in #2379 (Probably not that much...).

**Comment by Borda** [comment]
@sciunto correct, this patch does not cover #2379

File: skimage/measure/fit.py [hunk 1]
```diff
@@ -479,14 +490,18 @@ class EllipseModel(BaseModel):
...
+    def estimate(self, data, init_params=None):
         """Estimate circle model from data using total least squares.
 
         Parameters
         ----------
         data : (N, 2) array
             N points with ``(x, y)`` coordinates, respectively.
 
+        init_params : tuple of 5 values, optional
+            (x_center, y_center, a, b, theta) is initial guess of parameters.
```
Comment (sciunto) [inline skimage/measure/fit.py]: same here.

File: skimage/measure/fit.py [hunk 2]
```diff
@@ -479,14 +490,18 @@ class EllipseModel(BaseModel):
...
+            If they are not specified initial guess use a circle model.
```
Comment (sciunto) [inline skimage/measure/fit.py]: and here, as above.

File: skimage/measure/fit.py [hunk 3]
```diff
@@ -540,13 +555,19 @@ def Dfun(params):
 
             return A
 
-        # initial guess of parameters using a circle model
+        if init_params is None:
+            # initial guess of parameters using a circle model
+            xc0 = x.mean()
+            yc0 = y.mean()
+            r0 = np.sqrt((x - xc0) ** 2 + (y - yc0) ** 2).mean()
```
Comment (sciunto) [inline skimage/measure/fit.py]: Why don't you use the function `dist` here?

File: skimage/measure/fit.py [hunk 4]
```diff
@@ -737,8 +758,9 @@ def ransac(data, model_class, min_samples, residual_threshold,
 
         where `success` indicates whether the model estimation succeeded
         (`True` or `None` for success, `False` for failure).
-    min_samples : int
+    min_samples : int or float
         The minimum number of data points to fit a model to.
+        If the number is float in range [0, 1] it set the percentage of all data
```
Comment (sciunto) [inline skimage/measure/fit.py]: I would write: "If the number type is float, the value sets the percentage of all data."

File: skimage/measure/fit.py [hunk 5]
```diff
@@ -737,8 +758,9 @@ def ransac(data, model_class, min_samples, residual_threshold,
 
         where `success` indicates whether the model estimation succeeded
         (`True` or `None` for success, `False` for failure).
-    min_samples : int
+    min_samples : int or float
```
Comment (sciunto) [inline skimage/measure/fit.py]: or float in range [0, 1], optional

File: skimage/measure/fit.py [hunk 6]
```diff
@@ -858,8 +882,12 @@ def ransac(data, model_class, min_samples, residual_threshold,
     
     random_state = check_random_state(random_state)
 
-    if min_samples < 0:
+    if isinstance(min_samples, int) and min_samples < 0:
         raise ValueError("`min_samples` must be greater than zero")
+    elif isinstance(min_samples, float):
+        if not (0 < min_samples <= 1):
```
Comment (sciunto) [inline skimage/measure/fit.py]: Th inclusion or exclusion of zero is inconsistent with the docstring and error messages.

**Comment by sciunto** [comment]
In addition, it would be nice to have some test cases that cover the new options. :)

File: skimage/measure/fit.py (hunk 6 above)
Comment (Borda) [inline skimage/measure/fit.py]: not sure what you mean, the notation is too complex?

File: skimage/measure/fit.py (hunk 3 above)
Comment (Borda) [inline skimage/measure/fit.py]: here, dist is not defined as for Circle (only locally)

File: skimage/measure/fit.py (hunk 6 above)
Comment (sciunto) [inline skimage/measure/fit.py]: You should be consistent with the docstring. Is zero accepted or not?

**Comment by Borda** [comment]
is somewhere an explanation of codecov? because I am ot familiar with it and do not any idea how to correct code to pass this test...

**Comment by soupault** [comment]
@Borda oh, in the first approach it is quite intuitive.
Considering https://codecov.io/gh/scikit-image/scikit-image/compare/66743dc3c344b0f3c8bc60fd45d49ea465fe0784...acaf32fadd114bc194cdb2ff4c0c1a1e09591080 we have color indicator in the leftmost column which tells if the line affects the results of the tests.
I'd say that in order to make it all green you need to add couple of `assert_raises` catching those `ValueError`s.

**Comment by Borda** [comment]
it seem that continuous-integration/travis-ci/pr crash but in section with my code, how can I rerun it without committing any change? :)

**Comment by soupault** [comment]
@Borda I've restarted it.
You could create your own account in Travis CI, and integrate Borda/scikit-image there. This way you'lll definitely have the permissions required to manage the CI proccess.

**Comment by Borda** [comment]
@soupault thanks, but do you have an idea what is wrong with this doc generating
https://travis-ci.org/scikit-image/scikit-image/jobs/182044735

File: skimage/measure/fit.py [hunk 7]
```diff
@@ -479,14 +490,18 @@ class EllipseModel(BaseModel):
...
     """
 
-    def estimate(self, data):
+    def estimate(self, data, init_params=None):
         """Estimate circle model from data using total least squares.
 
         Parameters
         ----------
         data : (N, 2) array
             N points with ``(x, y)`` coordinates, respectively.
```
Comment (sciunto) [inline skimage/measure/fit.py]: Remove blank line

File: skimage/measure/fit.py [hunk 8]
```diff
@@ -341,14 +341,18 @@ class CircleModel(BaseModel):
...
```
Comment (sciunto) [inline skimage/measure/fit.py]: remove blank line

**Comment by soupault** [comment]
@Borda doesn't seem related to this PR. We have to adjust Sphinx config, I guess.

**Comment by sciunto** [comment]
I pushed a patch in #2395 let's see...

**Comment by Borda** [comment]
just rebase, no other change...

**Comment by Borda** [comment]
@sciunto are you fine with the code now? :)

**Comment by sciunto** [comment]
looks like your rebase is not correct. There is a mix of commits...

**Comment by Borda** [comment]
@sciunto I followed an advice you mentioned in #2394 - .github/CONTRIBUTING.txt
honestly, I am not familiar with the rebase, so maybe I made a mistake somewhere.
Just looking how to revert it back, not sure... [Undoing a git rebase - Stack Overflow](http://stackoverflow.com/questions/134882/undoing-a-git-rebase) ?

**Comment by sciunto** [comment]
Don't undo, try to rebase again. Your branch is not rebased and it seems that you will have some conflicts to solve.

**Comment by Borda** [comment]
@sciunto I passed the rebase again, but still it shows change over all other files (+2,272 lines −1,005 lines)

**Comment by sciunto** [comment]
I think you do something wrong each time you rebase. You have commits with the same message with different hashes... The better is to start a new PR on a new branch (either manually or if you are confident enough with git, cherry pick)

**Comment by Borda** [comment]
The last version was moved to #2415

**Pull request closed without merging**